design/batch_retrieval_design.md

```diff
@@ -1,20 +1,196 @@
+3. The old batch retrieval API accepts variable IDs as query parameters. We anticipate a long string of variableIDs to be passed in for secret retrieval and this impacts performance.
+4. Query parameters for HTTP GET requests are limited in length (2,048 characters) which is problematic if a customer has long secrets paths.
+5. The new route offers is more useful if we choose to expand its functionality in the future. By using POST instead of GET we can not only opt to use query parameters but also utilize the body of the request.
+
+At this stage this new batch retrieval API endpoint will be for internal use only.
+
+### Design
+
+#### Secrets Provider Job flow
+
+As a small recap, the following explains the Secrets Provider Job flow
+
+1. The `cyberark-secrets-provider-for-k8s` runs as a Job and authenticates with the Conjur/DAP Follower using the Kubernetes Authenticator (`authn-k8s`).
+
+2. The `cyberark-secrets-provider-for-k8s` reads all Kubernetes secrets required by the applications in the same namespace.
+
+3. For each mapped Kubernetes secret, 
+
+   a. The `cyberark-secrets-provider-for-k8s` performs a Batch Retrieval request and retrieves Conjur/DAP secrets.
+
+   b. It updates the Kubernetes secret with the Conjur/DAP secret value.
+
+4. The `cyberark-secrets-provider-for-k8s` Job is completed.
+
+The new Batch Retrieval API will effect #2a. 
+
+### URI
+
+##### POST /secrets/batch/v2
+
+This method retuns a list of secrets, their status code, and their value if an error did not occur. See difference between proposed API and existing API see [Proposed API vs Old API](#proposed-api-vs-old-api) below.
+
+```
+POST /secrets/batch/v2
+
+{
+  "variables": [
+    {"id": "var/path/conjurSecret"},
+    {"id": "var/path/anotherConjurSecret"},
+    ...
+  ]
+}
+```
+
+### Request Body
+
+The request body is a list of all variable IDs (paths to secrets in Conjur/DAP) that need to be retrieved. For example: 
+
+```
+{
+  "variables": [
+    {"id": "var/path/conjurSecret"},
+    {"id": "var/path/anotherConjurSecret"},
+    ...
+  ]
+}
+```
+
+#### Headers
+
+```
+Content-Type: application/json
+Authorization: Token token="eyJkYX...Rhb="
+```
+
+### Response
+
+#### Batch Response
+
+If no Conjur/DAP secrets were able to be retrieved (no 200 code), the response will still return a 200 OK because the [request body is parsable]([https://docs.microsoft.com/en-us/graph/json-batching#:~:text=The%20status%20code%20on%20a,requests%20inside%20the%20batch%20succeeded](https://docs.microsoft.com/en-us/graph/json-batching#:~:text=The status code on a,requests inside the batch succeeded)) and request was handled.
+
+If an invalid request body was supplied, a 400 Bad Request will be returned
+
+| Code            | Description                                                  |
+| --------------- | ------------------------------------------------------------ |
+| 200 OK          | All secret values were retrieved successfully                |
```

Comment:
What about to separate the use cases of "All secret values were retrieved successfully" and to "partial successes" , [here](https://stackoverflow.com/questions/8472935/http-status-code-for-a-partial-successful-request) is an example to the same issue, where they used `207 Multi-Status` response code for partial success

Reply:
What I gathered from [here](https://docs.microsoft.com/en-us/graph/json-batching#response-format) is that there are two types of status for the overall batch request 200 or 400. 
200 is always returned even if an internal response was a failure. 200 is returned if the body of the request is parsable. 
Do you still feel that a partial status, non-200 status code should be returned?

Reply:
Also found some good explanations here from Arik's blog post https://conjurhq.slack.com/files/U5VEFK01G/F01680UV11C/blog_-_how__and_why__to_build_killer_bulk_apis_-_part_2.pdf

Reply:
I think its OK to return 200 in the body of the request in case of partial failure